### PR TITLE
Remainings fixes before Botan runs fine in debug mode

### PR DESCRIPTION
**Pull request by webmaster128**
In this PR I am collecting manual fixes I made in order to make Botan in debug mode for #125.

Tested configurations (I'll add some more over time):

* `./botan-test` using `_GLIBCXX_DEBUG`
* Kullo using MSVC Debug library

### CODE REVIEW

**Comment by randombit** [comment]
Serpent change seems ok, but why does existing version trigger under a checker? W is already sized correctly.

**Comment by webmaster128** [comment]
Because if W is of length 140, the last valid element is W[139].



Mobil gesendet.

**Comment by randombit** [comment]
Thanks, I had missed that.

**Comment by randombit** [comment]
I'm not a huge fan of the 

`if(len > 0) { copy_mem(..., len); }`

changes. If these are needed can we revert that and hoist the check to `copy_mem` itself instead?

File: src/lib/mac/poly1305/poly1305.cpp [hunk 1]
```diff
@@ -190,7 +190,11 @@ void Poly1305::final_result(byte out[])
    if(m_buf_pos != 0)
       {
       m_buf[m_buf_pos] = 1;
-      clear_mem(&m_buf[m_buf_pos+1], m_buf.size() - m_buf_pos - 1);
+      const auto len = m_buf.size() - m_buf_pos - 1;
```
Comment (randombit) [inline src/lib/mac/poly1305/poly1305.cpp]: s/auto/size_t/ since that (I hope!) is what the compiler will deduce, and it makes for easier reading.

**Comment by webmaster128** [comment]
I guess in the majority if cases, moving the check into `copy_mem` would not work, because the bounds-check fails when evaluating the arguments of the `copy_mem` call.

E.g.

```
   if (in.second > 0)
       {
       copy_mem(&out[copy_offset], in.first, in.second);
       }
```
Now bounds-check is great in `&out[copy_offset]` because it ensures that `copy_offset` is somewhere in the middle if `out` when we copy. But it happens that `copy_offset` equals the length of out, i.e. beyond the vector.

This is in correct code only happening when we write 0 elements. If we write something >0 to a place beyond the array, the debug STL will crash, which is very valuable.

One could use `out.data() + copy_offset` but then we do crazy pointer arithmetics and loose the bounds-check.

File: src/lib/mac/poly1305/poly1305.cpp (hunk 1 above)
Comment (webmaster128) [inline src/lib/mac/poly1305/poly1305.cpp]: Done in #161

**Pull request merged**